Add Address Format

nerolation opened the pull request.

nerolation committed:
Add Address Format

nerolation committed:
Add scheme ID part

nerolation commented:
yo @mds1 , as soon as you're fine with the newly added address format + the scheme id location, we can push it to the ethereum/eips repo.

nerolation committed:
Update eip-5564.md

nerolation committed:
add bytes for metadata

mds1 commented on eip-5564.md:
```diff
@@ -1,7 +1,7 @@
-title: Non-Interactive Stealth Address Generation
-description: Stealth addresses for private transfers
+title: Stealth Addresses
+description: Stealth addresses for more private interaction
```
Thoughts on something like this?
```suggestion
description: Private, non-interactive transfers and interactions
```

nerolation replied:
Yeah I like that!

mds1 commented on eip-5564.md:
```diff
@@ -84,7 +84,7 @@ function computeStealthKey(
 The implementation of these methods is scheme-specific. The specification of a new stealth address scheme MUST specify the implementation for each of these methods. Additionally, although these function interfaces are specified in Solidity, they do not necessarily ever need to be implemented in Solidity, but any library or SDK conforming to this specification MUST implement these methods with com ...
 
-A one byte integer is used to identify stealth address schemes. A mapping from the scheme ID to it's specification is maintained in **{TODO where to maintain it?}**. An EIP is required to standardize a new stealth address scheme. These EIP extensions MUST specify:
+A one byte integer (`schemeID`) is used to identify stealth address schemes. The `schemeID` represents an incrementing integer beginning at 0. A mapping from the `schemeID` to it's specification MUST be declared in the EIP that proposes to standardize a new stealth address scheme. These EIP extensions MUST specify:
```
Nit, I personally prefer `schemeId` over `schemeID` (there's a bunch of places to change this if you're ok with the change)

nerolation replied:
Yeah I agree, I'll change it to schemeId.

mds1 commented on eip-5564.md:
```diff
@@ -99,23 +99,33 @@ This specification additionally defines a singleton `ERC5564Messenger` contract
   /// @param ephemeralPubKey Ephemeral public key used by the sender.
   /// @param metadata An arbitrary field that the sender can use however they like, but the below
   /// guidelines are recommended:
-  ///   - The first byte of the metadata MUST be the stealth address scheme ID.
-  ///   - The second byte of the metadata MUST be the view tag.
+  ///   - The first byte of the metadata MUST be the view tag.
   ///   - When sending ERC-20 tokens, the metadata SHOULD include the token address as the next 20
-  ///     bytes, and the amount being sent in the remaining 10 bytes.
+  ///     bytes, and the amount being sent in the remaining 11 bytes.
   ///   - When sending ERC-721 tokens, the metadata SHOULD include the token address as the first 20
-  ///     bytes, and the token ID being sent as remaining 10 bytes.
-  function announce(address stealthAddress, bytes memory ephemeralPubKey, bytes32 metadata)
+  ///     bytes, and the token ID being sent as remaining 11 bytes.
```
Now that we have more space, what do you think of this instead?

- The first byte of the metadata MUST be the view tag.
- When sending ERC-20 tokens, the metadata SHOULD be structured as follows:
    - Byte 1 MUST be the view tag, as specified above.
    - Bytes 2-5 are `23b872dd`, which the signature for the ERC-20 `transferFrom` method.
    - Bytes 6-25 are the token address.
    - Bytes 26-57 are the amount of tokens being sent.
- When approving a stealth address to spend ERC-20 tokens, the metadata SHOULD be structured as follows:
    - Byte 1 MUST be the view tag, as specified above.
    - Bytes 2-5 are `0xe1f21c67`, which the signature for the ERC-20 `approve` method.
    - Bytes 6-25 are the token address.
    - Bytes 26-57 are the approval amount.
- When sending ERC-721 tokens, the metadata SHOULD be structured as follows:
    - Byte 1 MUST be the view tag, as specified above.
    - Bytes 2-5 are `23b872dd`, which the signature for the ERC-721 `transfer` method.
    - Bytes 6-25 are the token address.
    - Bytes 26-57 are the token ID of the token being sent.

nerolation replied:
I think that 2-5 bytes it an overkill for the method id because it's already a the hashed canonical function and therefore quite unique. I think that 1 byte is more than enough, 2 at max because realistically, it's either a standard eth transaction or a transfer in 99% of the time and if it's an approval or another interaction, 1 byte should be sill enough to distinguish that.
On the other hand, what benefits do we provide an application if it knows, without looking into the contract or performing rpc calls, which exact function that was executed in a stealth interaction. F.e. if it's an approval, then the announcement event would already provide enough info for the app to suggest to execute an TransferFrom call.

In general, I really like the idea and think it makes the announcement much more valuable, though, we're not saving gas at all. Do you think we can use that much space, even though it is never stored and just temorary info emitted in an event?

mds1 replied:
> I think that 2-5 bytes it an overkill for the method id because it's already a the hashed canonical function and therefore quite unique. I think that 1 byte is more than enough, 2 at max because realistically, it's either a standard eth transaction or a transfer in 99% of the time and if it's an approval or another interaction, 1 byte should be sill enough to distinguish that.

I agree with all of this, in theory 1-2 bytes should be fine.

However, my thinking was that since we're trying to future-proof this standard (e.g. by using `bytes memory` for private keys instead of `uint256`), it seems we should be consistent and future proof here too—you never know what future ERCs/new (token) standards might emerge and have a 1-2 byte clash with approve or transferFrom. The cost of emitting a few extra bytes is quite small, so it seems like an ok tradeoff to me

mds1 commented on eip-5564.md:
```diff
@@ -99,23 +99,33 @@ This specification additionally defines a singleton `ERC5564Messenger` contract
+    uint256 schemeID, 
+    address stealthAddress, 
+    bytes memory ephemeralPubKey, 
+    bytes metadata
```
```suggestion
    bytes memory metadata
```

nerolation committed:
Update eip-5564.md

Co-authored-by: Matt Solomon <matt@mattsolomon.dev>

nerolation committed:
Update eip-5564.md

Co-authored-by: Matt Solomon <matt@mattsolomon.dev>

nerolation committed:
Update eip-5564.md

nerolation commented:
@mds1 btw, I've once mentioned that the messenger contract might have 1. either a function to route ETH through, allowing users to do stealth transfers of ETH in a single txs or 2. even have a very flexible fallback function even execute more complex functions, just like a proxie contract.
What do you think? 
Especially the ETH through-routing makes a lot of sense to me.

mds1 commented:
I don't think that needs to be in the messenger contract though, e.g. you can build a contract like this:

```solidity
contract SendEth {
  function sendEthToStealthAddress(...) external payable {
    Messenger.announce(...); // Emit event.
    stealthAddress.call{value: msg.value}(""); // Forward ETH to stealth address.
  }
}

nerolation commented:
That's true, though, wouldn't you agree that such a function would provide some value for the everyday EOA user. If users would like to try out stealth addresses, they might not want to deploy a contract for that purpose but still execute all in one txs.
It wouldn't add much complexity.

mds1 commented:
They wouldn't have to deploy a contract, as they can use someone else's! For example, we would almost certainly deploy a new version of [`Umbra.sol`](https://github.com/ScopeLift/umbra-protocol/blob/50bb314436c3a5b33e13ab50bcbce99c8a21e093/contracts-core/contracts/Umbra.sol#L118), and replace that linked line with a call to the Messenger contract

nerolation commented:
I got it! Yeah, makes fully sense!

nerolation committed:
Add Scheme Id mapping + update metadata

nerolation commented:
@mds1 I've updated the metadata as you suggested and added a sentence on the location of the scheme Id mapping.
I guess it's important to make sure we are future proof, though, as long as there're no quantumn schemes, the secp256k1 scheme may be the only one needed, therefore I'd say it's enough to specify it simply in an .md file that we put in the the ethereum/EIPs/assets/eip5564 folder so that everyone can access it.
Let me know what you think about the whole doc now. If you're fine with it, then I'll merge it into the ethereum/EIPs repo.
